### PR TITLE
paxos: skip logm op while trim reached max

### PR DESCRIPTION
**Pull request by YiteGu**
A large number of logm op will lead to continuous increase in mon store.db capacity. If the speed of trim cannot keep up with the speed of mon accepting logm op, mon db will write to full disk space.

Fixes: https://tracker.ceph.com/issues/66562

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by YiteGu** [comment]
@Matan-B have a code review :)

**Comment by YiteGu** [comment]
jenkins test make check

File: src/mon/PaxosService.h [hunk 1]
```diff
@@ -852,6 +857,24 @@ class PaxosService {
...
+  version_t get_last_to_remove() const {
+    return last_to_remove;
+  }
+
+  /**
+   * Check if the store.db disk space is insufficient 
+   *
+   * @returns true if insufficient space; false otherwise
+   * */
+  virtual bool is_insufficient_space() const {
```
[the remark this answers is not in the document]
Comment (YiteGu) [reply src/mon/PaxosService.h:874]: Got it. At first, I wanted HealthMonitor to rewrite this method, but now it unnecessary

File: src/mon/LogMonitor.cc [hunk 2]
```diff
@@ -699,7 +703,17 @@ bool LogMonitor::preprocess_log(MonOpRequestRef op)
...
-  
+
+  if (last_to_remove >= trim_max * trim_max_multiplier) {
+    dout(10) << "last trim quantities reached max, skip this op to avoid crash" << dendl;
+    goto done;
+  }
+
+  if (is_insufficient_space) {
+    dout(10) << "mon store.db space is insufficient, skip this op to avoid crash" << dendl;
+    goto done;
```
[the remark this answers is not in the document]
Comment (YiteGu) [reply src/mon/LogMonitor.cc:714]: Yes, I also think this requires a broader discussion. Matan and I discussed why Mon needs to store records logm, for example:
image
we think these records can ignore.

[the remark this answers is not in the document]
Comment (YiteGu) [reply src/mon/LogMonitor.cc:714]: > Am I understanding correctly that the jump to done below inhibits sending MLogAck, and thus the entries accumulate on senders?

Yes, no reply `MLogAck` to senders, the entries will accumulate in log_queue of senders.

[the remark this answers is not in the document]
Comment (YiteGu) [reply src/mon/LogMonitor.cc]: Still keep reply `MLogAck`

[the remark this answers is not in the document]
Comment (rzarzynski) [reply src/mon/LogMonitor.cc]: Not sending `MLogAck` is what makes sense, IMHO. That way we don't violate the contract on the log's persistence; instead an acute monitor failure becomes more gradual with cluster log failing (but not losing data yet) before other mon's subsystems.

[the remark this answers is not in the document]
Comment (YiteGu) [reply src/mon/LogMonitor.cc]: thanks for Radoslaw guidance.

**Comment by YiteGu** [comment]
jenkins test make check

**Comment by YiteGu** [comment]
jenkins test api

File: src/mon/LogMonitor.cc [hunk 3]
```diff
@@ -690,6 +690,10 @@ bool LogMonitor::preprocess_log(MonOpRequestRef op)
   auto m = op->get_req<MLog>();
   dout(10) << "preprocess_log " << *m << " from " << m->get_orig_source() << dendl;
   int num_new = 0;
+  const version_t trim_max = g_conf().get_val<version_t>("paxos_service_trim_max");
+  const uint64_t trim_max_multiplier = g_conf().get_val<uint64_t>("paxos_service_trim_max_multiplier");
+  const version_t last_to_remove = get_last_to_remove();
+  const bool is_insufficient_space = mon.paxos_service[PAXOS_HEALTH]->is_insufficient_space();
```
Comment (Matan-B) [inline src/mon/LogMonitor.cc:696]: is_insufficient_space can be used directly, right?
```
const bool is_insufficient_space = is_insufficient_space();
```

File: src/mon/PaxosService.h [hunk 4]
```diff
@@ -852,6 +857,24 @@ class PaxosService {
...
+  version_t get_last_to_remove() const {
+    return last_to_remove;
+  }
+
+  /**
+   * Check if the store.db disk space is insufficient
+   *
+   * @returns true if insufficient space; false otherwise
+   * */
+  bool is_insufficient_space() const {
```
Comment (Matan-B) [inline src/mon/PaxosService.h:874]: We can return const ref here and avoid the additional initialization in `LogMonitor::preprocess_log`

File: src/mon/PaxosService.h [hunk 5]
```diff
@@ -852,6 +857,24 @@ class PaxosService {
     return cached_last_committed;
   }
 
+  /**
+   * Get the quantity of last to remove
+   *
+   * @returns Our last to remove quantity
+   */
+  version_t get_last_to_remove() const {
```
Comment (Matan-B) [inline src/mon/PaxosService.h:865]: ditto

File: src/mon/PaxosService.h (hunk 4 above)
Comment (Matan-B) [reply src/mon/PaxosService.h:874]: Either move to protected

File: src/mon/LogMonitor.cc [hunk 6]
```diff
@@ -699,7 +703,17 @@ bool LogMonitor::preprocess_log(MonOpRequestRef op)
 	    << session->caps << dendl;
     goto done;
   }
-  
+
+  if (last_to_remove >= trim_max * trim_max_multiplier) {
+    dout(10) << "last trim quantities reached max, ignore this op to avoid db growing too quickly" << dendl;
+    goto done;
+  }
```
[the remark this answers is not in the document]
Comment (YiteGu) [reply src/mon/LogMonitor.cc]: Yes, the maximum value of `last_to_remove` can only be `trim_max * trim_max_multiplier`
updated

File: src/mon/Monitor.cc [hunk 7]
```diff
@@ -5961,6 +5961,9 @@ void Monitor::tick()
   }
 
   mgr_client.update_daemon_health(get_health_metrics());
+
+  const utime_t lat = ceph_clock_now() - now;
+  dout(11) << " tick completed, latency: " << lat << dendl;
```
[the remark this answers is not in the document]
Comment (YiteGu) [reply src/mon/Monitor.cc]: I got it

File: src/mon/PaxosService.h (hunk 5 above)
Comment (YiteGu) [reply src/mon/PaxosService.h]: done

**Comment by YiteGu** [comment]
@Matan-B 
1. I have cancelled is_insufficient_space, we cannot accurately determine the absolute value of the remaining space on the user's mon disk.
2. We can be certain that if `last_to_remove` is equal `trim_max * trim_max_multiplier`, we can completely let mon temporarily ignore logm until the trim pressure decreases.

**Comment by YiteGu** [comment]
jenkins test make check

File: src/mon/LogMonitor.cc [hunk 8]
```diff
@@ -699,7 +702,12 @@ bool LogMonitor::preprocess_log(MonOpRequestRef op)
 	    << session->caps << dendl;
     goto done;
   }
-  
+
+  if (last_to_remove == trim_max * trim_max_multiplier) {
+    dout(10) << "last trim quantities reached max, ignore this op to avoid db growing too quickly" << dendl;
```
Comment (Matan-B) [inline src/mon/LogMonitor.cc:707]: nit: line length

File: src/mon/LogMonitor.cc [hunk 9]
```diff
@@ -699,7 +702,12 @@ bool LogMonitor::preprocess_log(MonOpRequestRef op)
 	    << session->caps << dendl;
     goto done;
   }
-  
+
+  if (last_to_remove == trim_max * trim_max_multiplier) {
```
Comment (Matan-B) [inline src/mon/LogMonitor.cc:706]: We should check that `trim_max_multiplier` actually has a value. if `trim_max_multiplier` equals to zero trim max is disabled and we would ignore all cases in which `last_to_remove` is zero.
```
if (trim_max_multiplier && last_to_remove == trim_max * trim_max_multiplier)
```
should probably do.

**Review by Matan-B — CHANGES_REQUESTED** [review]
lgtm, one comment.

File: src/mon/LogMonitor.cc (hunk 8 above)
Comment (YiteGu) [reply src/mon/LogMonitor.cc]: done

File: src/mon/LogMonitor.cc (hunk 9 above)
Comment (YiteGu) [reply src/mon/LogMonitor.cc]: done

**Comment by YiteGu** [comment]
jenkins test make check

**Pull request closed without merging**